### ISSUE TITLE
Settle on an interface for the "Driver" behaviour

### ISSUE THREAD

**Issue by gausby** [gausby/tortoise#4]
When event happen (such as the client receiving a message) a callback will get called in a user defined callback module. This module has to implement the `Tortoise.Driver` behaviour.

Currently the events triggered are when a ping response is received (allow the user to log the duration of the ping somehow), and when one of the subscribed topics receive a message. We need to specify which events are interesting for the end-user, and the names for the callback functions. See the `Tortoise.Driver.Logger` module for an implementation of a driver.

Subject for discussion:

- What events are interesting?
- What should the callbacks be named?

**Comment by gausby** [comment]
I think I will remove the `ping_response/2` callback from the driver. The thought was that one could use this to log the latency to the server, but I think there are much nicer ways of doing that—ways that doesn't require the user to concern themselves with this detail about the MQTT protocol(1), and removing it will mean one less callback for the user to implement.

____
(1) The ping-request/ping-response is used for keeping the connection alive, if no messages send from the client to the server in the specified Keep-alive interval the server will drop the connection. The `Tortoise.Connection` process will ensure we send a ping request so we adhere to the keep alive.

**Comment by fhunleth** [comment]
Regarding `ping_response/2` and `disconnect/1`, we need to know when the client is connected to the server to update a status LED (and possibly behave differently). I don't think we would do much with `ping_response/2` unless we needed it to derive the connection state.

**Comment by gausby** [comment]
I think I will go with the following callbacks:

`connection/2` (on `connack`/`connection dropped`)
Triggered when the connection state changes. First argument is `:up` or `:down` and the second argument is the internal state that can be used by the implementor to keep state between messages.

`subscription/3` (on `suback`/`unsuback`)
Triggered when there is a change in the subscription state. The first argument will be either `:up` or `:down`, the second argument the topic filter that was subscribed to (or unsubscribed from) and the third argument is the internal state. 

`handle_message/3` (on `publish`)
Triggered when the client receive a message from the broker. The first argument is the topic string, which has been split into a list on forward slash (/), making it possible to pattern match on the topic using wild cards, so if we subscribe to the topic filter `foo/+/baz` we can pattern match on `["foo", "bar", "baz"]` and `["foo", _, "baz"]`, etc.

The second argument will be the message payload. I guess we will run into trouble when we start implementing the MQTT 5 protocol, because the message can carry mime-type information, in that case we could put the payload in a tuple: `{"application/json", payload}` but for now it will just be `payload`.

The third argument is the state the implementor can use to store state between messages.

Notice that we will not be concerned with the quality of service of the message. The end user should not be concerned with this implementation detail besides when they setup the subscription. We are only concerned with when the message changes owner (is received), and we are only concerned with incoming messages in this part of the system.

`terminating/2` 
Triggered when the connection is being disconnected and closed down. First argument is `reason` and the second argument is the state.

**Comment by gausby** [comment]
Perhaps the Driver should be called Handler. Driver might be confused with the transport type ?

**Comment by gausby** [comment]
My reason I want the `:up` and `:down` arguments in the `subscription/3` callback is that life time logic for a given topic can be keep together.

```elixir
def subscription(:up, "foo/+/baz", state) do
  # ...
end

def subscription(:down, "foo/+/baz", state) do
  # ...
end

def subscription(:up, "hello/world", state) do
  # ...
end

def subscription(:down, "hello/world", state) do
  # ...
end
#... etc
```

**Comment by fhunleth** [comment]
What's the difference between `connection(:down, state)` and `terminating(_reason, state)`? Is the former temporary since `tortoise` will try to reconnect and the latter one permanent?

**Comment by gausby** [comment]
Yes. I see `terminating/2` as we sent the `Disconnect` message to the server, so we are terminating with a "normal"-reason. We can use terminating to close a database connection, or other stuff that we setup in the init.

Connection down will hint that we are trying to reestablish the connection—i.e. not yet about to drop the state.

**Comment by gausby** [comment]
In #15 I've implemented the `handle_message/3` and `subscription/3` callbacks. There are some open questions with regards to `subscription/3`

- What should happen when a different QoS is returned for the subscription than the one requested: i.e. we subscribe to *foo/bar qos 2* but is accepted with *foo/bar qos 1*. I guess we should warn about this somehow (right now the implementation I have in the pull request will crash because I only match on the happy case)

**Comment by gausby** [comment]
I've merged PR #15 this implements `init/1`, `connection/2`, `subscription/3`, `handle_message/3`, and `terminate/2`. I might open issues on problems and questions I've identified while implementing the callback module; and the `Driver` will be renamed to `Handler` eventually.
